clippy_lints/src/unnecessary_reserve.rs

```diff
@@ -0,0 +1,151 @@
+        if !meets_msrv(self.msrv, msrvs::UNNECESSARY_RESERVE) {
+            return;
+        }
+
+        for (idx, stmt) in block.stmts.iter().enumerate() {
+            if let StmtKind::Semi(semi_expr) = stmt.kind
+                && let ExprKind::MethodCall(_, [struct_calling_on, _], _) = semi_expr.kind
+                && let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(semi_expr.hir_id)
+                && (match_def_path(cx, expr_def_id, &paths::VEC_RESERVE) ||
+                    match_def_path(cx, expr_def_id, &paths::VEC_DEQUE_RESERVE))
+                && acceptable_type(cx, struct_calling_on)
+                && let Some(next_stmt_span) = check_extend_method(cx, block, idx, struct_calling_on)
+                && !next_stmt_span.from_expansion()
+            {
+                span_lint_and_then(
+                    cx,
+                    UNNECESSARY_RESERVE,
+                    next_stmt_span,
+                    "unnecessary call to `reserve`",
+                    |diag| {
+                        diag.span_suggestion(
+                            semi_expr.span,
+                            "remove this line",
+                            String::new(),
+                            Applicability::MaybeIncorrect,
+                        );
+                    }
+                );
+            }
+        }
+    }
+
+    extract_msrv_attr!(LateContext);
+}
+
+#[must_use]
+fn acceptable_type(cx: &LateContext<'_>, struct_calling_on: &rustc_hir::Expr<'_>) -> bool {
+    let acceptable_types = [sym::Vec, sym::VecDeque];
+    acceptable_types.iter().any(|&acceptable_ty| {
+        match cx.typeck_results().expr_ty(struct_calling_on).peel_refs().kind() {
+            ty::Adt(def, _) => cx.tcx.is_diagnostic_item(acceptable_ty, def.did()),
+            _ => false,
+        }
+    })
+}
+
+#[must_use]
+fn check_extend_method(
+    cx: &LateContext<'_>,
+    block: &Block<'_>,
+    idx: usize,
+    struct_expr: &rustc_hir::Expr<'_>,
+) -> Option<rustc_span::Span> {
+    let mut read_found = false;
+    let next_stmt_span;
+
+    let mut visitor = expr_visitor_no_bodies(|expr| {
+        if let ExprKind::MethodCall(_, [struct_calling_on, _], _) = expr.kind
+            && let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id)
+            && match_def_path(cx, expr_def_id, &paths::ITER_EXTEND)
+            && acceptable_type(cx, struct_calling_on)
+            && equal_ident(struct_calling_on, struct_expr)
+        {
+            read_found = true;
+        }
+        !read_found
+    });
+
+    if idx == block.stmts.len() - 1 {
+        if let Some(e) = block.expr {
+            visitor.visit_expr(e);
+            next_stmt_span = e.span;
+        } else {
+            return None;
+        }
+    } else {
+        let next_stmt = &block.stmts[idx + 1];
+        visitor.visit_stmt(next_stmt);
+        next_stmt_span = next_stmt.span;
+    }
+    drop(visitor);
```

Comment:
Why? :thinking:

Reply:
`read_found` was used in the closure, and need to be checked.

Reply:
Ah ok. I think the refactor in #8762 will make this unnecessary. Let's see what will get merged first.